pkg/controller/addon/addon_controller.go

```diff
@@ -0,0 +1,358 @@
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+	addons map[string]mf.Manifest
+}
+
+// Reconcile reads that state of the cluster for a Addon object and makes changes based on the state read
+// and what is in the Addon.Spec
+func (r *ReconcileAddon) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := requestLogger(req, "addon reconcile")
+	//reqLogger := log.WithValues("Request.Namespace", req.Namespace, "Request.Name", req.Name)
+	log.Info("Reconciling Addon")
+
+	// Fetch the Addon instance
+	instance := &op.Addon{}
+	err := r.client.Get(context.TODO(), req.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			log.Info("resource has been deleted")
+			return reconcile.Result{}, nil
+			// addon components (items in yaml manifest) will be deleted
+			// as the owner reference is set, technically we don't have to explicitly delete them
+			// if we wan't to delete them manually (if in case of orphaned items)
+			// a more complex reconcile logic can be implemented here
+			// for example: 1. read manifest, 2. find pipeline-installation,
+			// 3. note: can't set owner reference as the addon resource is not found
+			// 4. inject namespace into resources suing target namespace from pipeline
+			// 5. delete all resources
+			// NT: aug-22-2019
+		}
+		return reconcile.Result{}, err
+	}
+
+	//if no version is specified in spec
+	//then, try to set latest available version for the addon in addon spec
+	ok, err := r.ensureAddonVersion(instance)
+
+	if !ok {
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	if isAddOnUpToDate(instance) {
+		log.Info("skipping installation, resource already up to date")
+		return reconcile.Result{}, nil
+	}
+
+	return r.reconcileAddon(req, instance)
+}
+
+func (r *ReconcileAddon) ensureAddonVersion(res *op.Addon) (bool, error) {
+	version := res.Spec.Version
+	if version != "" {
+		return true, nil
+	}
+	version, err := GetLatestVersion(res)
+	if err != nil {
+		return false, err
+	}
+
+	tmpRes := res.DeepCopy()
+	tmpRes.Spec.Version = version
+	err = r.client.Update(context.TODO(), tmpRes)
+	if err != nil {
+		return false, err
+	}
+	r.refreshCR(res)
+
+	return true, nil
+}
+
+func isAddOnUpToDate(res *op.Addon) bool {
+	c := res.Status.Conditions
+	if len(c) == 0 {
+		return false
+	}
+	latest := c[0]
+	return latest.Version == res.Spec.Version &&
+		latest.Code == op.InstalledStatus
+}
+
+func (r *ReconcileAddon) reconcileAddon(req reconcile.Request, res *op.Addon) (reconcile.Result, error) {
+	log := requestLogger(req, "addon install")
+
+	err := r.updateStatus(res, op.AddonCondition{Code: op.InstallingStatus, Version: res.Spec.Version})
+	if err != nil {
+		log.Error(err, "failed to set status")
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	//find the valid clusterwide tekton-pipeline installation
+	piplnRes, err := r.pipelineReady()
```

Comment:
how about below to improve readabilty? 

```go
config, err := r.getClusterConfig()
if err {
...
}
if err := r.validateClusterConfig(config); err != nil {
   .. 
}
```

Reply:
ok

Reply:
i prefer to use the  word `pipeline` in this context. We might know what we mean by `cluster - config` here. But it is a phrase widely used in Kubernetes ecosystem to denote something related to the cluster itself.